src/collision.c

```diff
@@ -0,0 +1,1084 @@
+                if (test_ray_triangle_general(triangle++, pCollisionData->vertices))
+                    colliderID = i;
+        }
+    }
+
+    if (colliderID >= 0) {
+        *hitX = gCollisionPointX;
+        *hitY = gCollisionPointY;
+        *hitZ = gCollisionPointZ;
+        *hitDepth = gCollisionRayLength;
+        *hitNx = gCollisionNormalX;
+        *hitNy = gCollisionNormalY;
+        *hitNz = gCollisionNormalZ;
+        return colliderID;
+    } else
+        return colliderID;
+}
+
+s32 test_ray_zones(f32 startX, f32 startY, f32 startZ, f32 dirX, f32 dirY, f32 dirZ,
+                f32* hitX, f32* hitY, f32* hitZ, f32* hitDepth, f32* hitNx, f32* hitNy, f32* hitNz) {
+    Collider* collider;
+    CollisionData* pCollisionData;
+    ColliderTriangle* triangle;
+    s32 i, j;
+    s32 colliderID;
+
+    pCollisionData = &gZoneCollisionData;
+    gCollisionRayDirX = dirX;
+    gCollisionRayDirY = dirY;
+    gCollisionRayDirZ = dirZ;
+    gCollisionRayStartX = startX;
+    gCollisionRayStartY = startY;
+    gCollisionRayStartZ = startZ;
+    gCollisionRayLength = *hitDepth;
+    colliderID = -1;
+
+    for (i = 0; i < pCollisionData->numColliders; i++) {
+        collider = &pCollisionData->colliderList[i];
+
+        if (collider->flags & 0x10000)
+            continue;
+
+        if (collider->numTriangles == 0 || collider->aabb == NULL)
+            continue;
+
+        triangle = collider->triangleTable;
+        for (j = 0; j < collider->numTriangles; j++)
+            if (test_ray_triangle_down(triangle++, pCollisionData->vertices))
+                colliderID = i;
+    }
+
+    if (colliderID >= 0) {
+        *hitX = gCollisionPointX;
+        *hitY = gCollisionPointY;
+        *hitZ = gCollisionPointZ;
+        *hitDepth = gCollisionRayLength;
+        *hitNx = gCollisionNormalX;
+        *hitNy = gCollisionNormalY;
+        *hitNz = gCollisionNormalZ;
+        return colliderID;
+    } else
+        return colliderID;
+}
+
+f32 test_ray_collider_horizontal(s32 ignoreFlags, s32 colliderID, f32 x, f32 y, f32 z, f32 length, f32 yaw) {
+    CollisionData* collisionData = &gCollisionData;
+    f32 cosTheta;
+    f32 sinTheta;
+    Collider* collider;
+    ColliderTriangle* triangleTable;
+    s32 i;
+    f32 ret;
+
+    sin_cos_rad(yaw * TAU / 360.0f, &sinTheta, &cosTheta);
+    collider = &collisionData->colliderList[colliderID];
+
+    gCollisionRayDirY = 0;
+    gCollisionRayStartX = x;
+    gCollisionRayStartY = y;
+    gCollisionRayStartZ = z;
+    gCollisionRayLength = length;
+    gCollisionRayDirX = sinTheta;
+    gCollisionRayDirZ = -cosTheta;
+    ret = -1.0f;
+
+    if (!(collider->flags & ignoreFlags) && collider->numTriangles != 0) {
+        triangleTable = collider->triangleTable;
+
+        for (i = 0; i < collider->numTriangles; i++)
+            if (test_ray_triangle_horizontal(triangleTable++, collisionData->vertices))
+                ret = gCollisionRayLength;
+    }
+
+    return ret;
+}
+
+typedef struct {
```

Comment:
This struct def and these externs should be at the top with the others (externs in ram order too please)